### PR TITLE
Split Out Primitive Architecture To Allow It To Run On x86 Systems

### LINKED ISSUE #790: Restructure Firmware
### Description of the task

We want to refactor and cleanup the firmware for both the robots and dongle. This is difficult since there are no tests, so integration tests are our only option to make sure things work.

We should figure out what areas / modules we want to refactor and in what order, and figure out how we will test along the way to make sure we don't break anything.

### Acceptance criteria

- [ ] Make a plan for firmware refactoring and testing

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Fully split the primitive architecture into the `app` layer of firmware, so that it can be completely compiled and used on a `x86` system. 

- Moved all primitives to `firmware/main/app/primitives`
- Ported all app-layer control stuff into the `app` folder
- Created a `PrimitiveManager` (C-style) class in `firmware/main/app/primitive_manager.h` to control the running of primitives and to hold primitive state
- Lots of cleanup of warnings that are now compiler errors
- Lots of misc. moving and renaming of stuff so that we can actually compile stuff independent of hardware specific things like FreeRTOS

### Testing Done
- Unit tests for new `app/world` stuff
- field tested on a robot, verified functionality works via MRF test

### Resolved Issues
resolves #790 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/cc_toolchain/BUILD [hunk 1]
```diff
@@ -23,6 +23,21 @@ environment_group(
...
+
+config_setting(
+    name = "cpu_stm32f4",
+    values = {"cpu": "stm32f4"},
+)
+
+config_setting(
+    name = "cpu_stm32h7",
+    values = {"cpu": "stm32h7"},
+)
```
Comment (garethellis0) [inline src/cc_toolchain/BUILD:39]: Unfortunately we need these to use the `select` in `firmware/main/app/primitives/BUILD`, we can't just use the `cpu` from the environment directly.

File: src/firmware/cdcacm/include/cdcacm.h [hunk 2]
```diff
@@ -16,7 +16,7 @@ typedef enum
 /**
  * \brief Descriptor subtypes used in CDC ACM devices.
  */
-typedef enum
+enum
```
Comment (garethellis0) [inline src/firmware/cdcacm/include/cdcacm.h:19]: fixed a compiler warning (now error) here.

File: src/firmware/main/app/primitives/primitive_manager.c [hunk 3]
```diff
@@ -0,0 +1,187 @@
+#include "primitive_manager.h"
+
+#ifdef __arm__
+// TODO: do we need "FreeRTOS.h" here?
```
Comment (MathewMacDougall) [inline src/firmware/main/app/primitives/primitive_manager.c]: TODO?

File: src/firmware/main/app/primitives/primitive_manager.c [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+#endif
+}
+
+PrimitiveManager_t *app_primitive_manager_create(void)
+{
+    PrimitiveManager_t *manager =
+        (PrimitiveManager_t *)malloc(sizeof(PrimitiveManager_t));
+
+#ifdef __arm__
+    // TODO: this could cause issues if we ever have multiple PrimitiveManager's running
```
Comment (MathewMacDougall) [inline src/firmware/main/app/primitives/primitive_manager.c]: TODO?

Comment (MathewMacDougall) [reply src/firmware/main/app/primitives/primitive_manager.c]: Also I don't think this is a concern we already made tons of assumptions about only one instance of stuff existing at once in firmware since it's per-robot

File: src/firmware/main/app/primitives/primitive_manager.c [hunk 5]
```diff
@@ -0,0 +1,187 @@
+#include "primitive_manager.h"
+
+#ifdef __arm__
+// TODO: do we need "FreeRTOS.h" here?
+#include <FreeRTOS.h>
+#include <semphr.h>
+#elif __unix__
```
Comment (MathewMacDougall) [inline src/firmware/main/app/primitives/primitive_manager.c:8]: Maybe a quick comment for this file explaining why the typedefs are needed would be useful

File: src/firmware/main/app/primitives/shoot.c [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+typedef struct ShootPrimitiveState
+{
+    float destination[3], major_vec[2], minor_vec[2], total_rot;
+} ShootPrimitiveState_t;
+DEFINE_PRIMITIVE_STATE_CREATE_AND_DESTROY_FUNCTIONS(ShootPrimitiveState_t)
+
+
+/**
+ * Scales the major acceleration by the distance from the major axis and the
+ * amount required left to rotate. Total roation and the distance vector should
```
Comment (MathewMacDougall) [inline src/firmware/main/app/primitives/shoot.c]:
```suggestion
 * amount required left to rotate. Total rotation and the distance vector should
```

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/firmware/main/app/control/control.h [hunk 7]
```diff
@@ -1,6 +1,8 @@
 #pragma once
 
-#include "app/world/firmware_robot.h"
+#include "firmware/main/app/world/firmware_robot.h"
+
+// TODO: need to prefix everything with "_app"
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/firmware/main/app/control/control.h]: done

File: src/firmware/main/app/control/physbot.c [hunk 8]
```diff
@@ -0,0 +1,52 @@
+#include "firmware/main/app/control/physbot.h"
+
+#include "firmware/main/app/control/bangbang.h"
+#include "firmware/main/shared/physics.h"
+
+// TODO: this is redefined in so many places, should be centralized?
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/firmware/main/app/control/physbot.c]: This falls under the umbrella of a ton of small fixes that need to be made. Given that the control stuff is going to be replaced soon, I don't think it's worth fixing now.

File: src/firmware/main/app/primitives/primitive_manager.c (hunk 3 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/primitive_manager.c]: we do, removed comment

File: src/firmware/main/app/primitives/primitive_manager.c (hunk 4 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/primitive_manager.c]: Yeah, just wanted to come back and look at this, but I agree for now.

File: src/firmware/main/app/primitives/primitive_manager.c (hunk 5 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/primitive_manager.c:8]: added:
```
// There are different semaphore implementations depending on if we're on a x86 or arm
// system, and we use typdefs here to switch between them
```

File: src/firmware/main/app/primitives/shoot.c (hunk 6 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/shoot.c]: fixed

**Comment by MathewMacDougall** [comment]
Nothing more to add pending a field test and CI. Seems like CI is failing due to the `select` changes in bazel?

@EvanMorcom @jonathanlew @akhilveeraghanta @LiCody @amrm3lm Could use a review ASAP so we can try get this merged by Saturday's meeting.

File: src/firmware/main/app/control/physbot.h [hunk 9]
```diff
@@ -71,7 +67,7 @@ PhysBot setup_bot(dr_data_t states, float destination[3], float major_vec[2],
  * @param p A 3 length array of {final velocity, max acceleration, max_velocity}
  * @return void
  */
-void plan_move(Component *c, float p[3]);
+void app_physbots_planMove(Component *c, float *p);
```
Comment (akhilveeraghanta) [inline src/firmware/main/app/control/physbot.h]: `app_physbots` -> `app_physbot`

File: src/firmware/main/app/primitives/imu_test_primitive.c [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    state->y_dest     = (float)(params->params[1] / 1000.0f);
+    state->avel_final = (float)(params->params[2] / 100.0f);
+    state->slow       = params->slow;
+}
+
+static void imu_test_end(void* void_state_ptr, FirmwareWorld_t* world) {}
+
+static void imu_test_tick(void* void_state_ptr, FirmwareWorld_t* world)
+{
+    // TODO: this primitive does nothing. Delete?
```
Comment (akhilveeraghanta) [inline src/firmware/main/app/primitives/imu_test_primitive.c:45]: can we nuke it now?

File: src/firmware/main/app/primitives/move_primitive.c [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+    (*wheel_axes)[1] = angle + ANGLE_TO_FRONT_WHEELS + PI_2;
+    (*wheel_axes)[2] = angle - ANGLE_TO_FRONT_WHEELS - PI_2;
+    (*wheel_axes)[3] = angle - ANGLE_TO_FRONT_WHEELS + PI_2;
+    (*wheel_axes)[4] = angle + ANGLE_TO_BACK_WHEELS - PI_2;
+    (*wheel_axes)[5] = angle + ANGLE_TO_BACK_WHEELS - (3 * PI_2);
+    (*wheel_axes)[6] = angle - ANGLE_TO_BACK_WHEELS + PI_2;
+    (*wheel_axes)[7] = angle - ANGLE_TO_BACK_WHEELS + (3 * PI_2);
+}
+
+// TODO: finish jdoc?
```
Comment (akhilveeraghanta) [inline src/firmware/main/app/primitives/move_primitive.c]: there already is a jdoc up top, can either nuke this or move that here

File: src/firmware/main/app/primitives/primitive_manager.h [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+                                             unsigned int primitive_index,
+                                             const primitive_params_t *params);
+
+/**
+ * Runs the current primitive
+ *
+ * @param manager The primitive manager to set the current primitive on
+ * @param world The world to run the primitive in
+ */
+void app_primitive_manager_run_current_primitive(PrimitiveManager_t *manager,
```
Comment (akhilveeraghanta) [inline src/firmware/main/app/primitives/primitive_manager.h]: `app_primitive_manager_run_current_primitive` -> `app_primitive_manager_runCurrentPrimitive`

**Review by akhilveeraghanta — DISMISSED** [review]
LG:tm: overall, just some minor nitpicks

File: src/firmware/main/app/primitives/catch_primitive.c [hunk 13]
```diff
@@ -170,21 +129,22 @@ static void catch_tick(log_record_t* log, FirmwareWorld_t* world)
...
         float max_major_v = 3.0;  //(get_var(0x01)/4.0);
         float major_vel   = major_vec[0] * vel[0] + major_vec[1] * vel[1];
-        PrepareBBTrajectoryMaxV(&major_profile, major_disp, major_vel, end_speed,
-                                max_major_a, max_major_v);  // 3.5, 3.0
-        PlanBBTrajectory(&major_profile);
-        major_accel      = BBComputeAvgAccel(&major_profile, TIME_HORIZON);
-        float time_major = GetBBTime(&major_profile);
+        app_bangbang_prepareTrajectoryMaxV(&major_profile, major_disp, major_vel,
+                                           end_speed, max_major_a,
+                                           max_major_v);  // 3.5, 3.0
```
Comment (EvanMorcom) [inline src/firmware/main/app/primitives/catch_primitive.c:134]: Any idea what the commented numbers are?

File: src/firmware/main/app/primitives/dribble_primitive.c [hunk 14]
```diff
@@ -0,0 +1,88 @@
+#include "firmware/main/app/primitives/dribble_primitive.h"
+
+#include <stdio.h>
+
+#include "firmware/main/app/control/bangbang.h"
+#include "firmware/main/app/control/control.h"
+#include "firmware/main/shared/physics.h"
+#include "firmware/main/shared/util.h"
+
+#define DRIBBLE_TIME_HORIZON 0.05f  // s
```
Comment (EvanMorcom) [inline src/firmware/main/app/primitives/dribble_primitive.c:10]: Same as above, maybe explain what the TIME_HORIZON is

File: src/firmware/main/app/primitives/imu_test_primitive.c (hunk 10 above)
Comment (EvanMorcom) [reply src/firmware/main/app/primitives/imu_test_primitive.c:45]: 💥

File: src/firmware/main/app/primitives/move_primitive.c [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+#include "firmware/main/app/control/control.h"
+#include "firmware/main/app/control/physbot.h"
+#include "firmware/main/shared/physics.h"
+#include "firmware/main/shared/util.h"
+#include "shared/constants.h"
+#include "shared/robot_constants.h"
+
+// these are set to decouple the 3 axis from each other
+// the idea is to clamp the maximum velocity and acceleration
+// so that the axes would never have to compete for resources
```
Comment (EvanMorcom) [inline src/firmware/main/app/primitives/move_primitive.c:16]: Comment not seeming related to TIME_HORIZON again

File: src/firmware/main/app/primitives/pivot_primitive.c [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
+typedef struct PivotPrimitiveState
+{
+    float radius, speed, angle, center[2], final_dest[2];
+    int dir;
+} PivotPrimitiveState_t;
+DEFINE_PRIMITIVE_STATE_CREATE_AND_DESTROY_FUNCTIONS(PivotPrimitiveState_t)
+
+/**
+ * \brief Does bangbang stuff and returns the acceleration value
```
Comment (EvanMorcom) [inline src/firmware/main/app/primitives/pivot_primitive.c:27]: Grade A description

File: src/firmware/main/app/primitives/primitive_manager.c [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+    app_dribbler_setSpeed(dribbler, 0);
+    manager->current_primitive       = PRIMITIVES[primitive_index];
+    manager->current_primitive_index = primitive_index;
+    manager->current_primitive_state = manager->current_primitive->create_state();
+    manager->current_primitive->start(params, manager->current_primitive_state, world);
+
+    app_primitive_manager_unlockPrimitiveMutex(manager);
+}
+
+void app_primitive_manager_run_current_primitive(PrimitiveManager_t *manager,
```
Comment (LiCody) [inline src/firmware/main/app/primitives/primitive_manager.c]:
```suggestion
void app_primitive_manager_runCurrentPrimitive(PrimitiveManager_t *manager,
```

**Review by EvanMorcom — CHANGES_REQUESTED** [review]
About halfway done. I will pick it up again later.

🔥

File: src/firmware/main/control/control.c [hunk 18]
```diff
@@ -2,12 +2,12 @@
 
 #include <math.h>
 
+#include "firmware/main/shared/physics.h"
 #include "io/adc.h"
 #include "io/dr.h"
 #include "io/dsp.h"
 #include "io/encoder.h"
 #include "io/wheels.h"
```
Comment (LiCody) [inline src/firmware/main/control/control.c:10]:
```suggestion
#include "firmware/main/io/wheels.h"
```

File: src/firmware/main/io/dr.c [hunk 19]
```diff
@@ -4,11 +4,10 @@
 #include <stdint.h>
 #include <stdio.h>
 
+#include "firmware/main/shared/physics.h"
 #include "io/dsp.h"
 #include "io/encoder.h"
 #include "io/sensors.h"
```
Comment (LiCody) [inline src/firmware/main/io/dr.c:10]: same as above
```suggestion
#include "firmware/main/io/sensors.h"
```

File: src/firmware/main/io/dribbler.c [hunk 20]
```diff
@@ -7,11 +7,11 @@
 #include <stdio.h>
 
 #include "control/control.h"
+#include "firmware/main/shared/physics.h"
 #include "io/adc.h"
 #include "io/hall.h"
 #include "io/motor.h"
 #include "io/receive.h"
```
Comment (LiCody) [inline src/firmware/main/io/dribbler.c:14]: same as above

File: src/firmware/main/io/receive.c [hunk 21]
```diff
@@ -33,8 +35,6 @@
 #include "io/motor.h"
 #include "io/mrf.h"
 #include "main.h"
```
Comment (LiCody) [inline src/firmware/main/io/receive.c:37]: same as above

File: src/firmware/main/main.c [hunk 22]
```diff
@@ -48,14 +49,14 @@
...
 #include "io/icb.h"
 #include "io/leds.h"
+#include "io/lps.h"
 #include "io/motor.h"
 #include "io/mrf.h"
 #include "io/pins.h"
 #include "io/receive.h"
 #include "io/sdcard.h"
 #include "io/usb_config.h"
 #include "io/wheels.h"
```
Comment (LiCody) [inline src/firmware/main/main.c:59]: same as above

**Review by LiCody — CHANGES_REQUESTED** [review]
LGTM mostly just small stuff

File: src/firmware/main/app/primitives/imu_test_primitive.c (hunk 10 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/imu_test_primitive.c:45]: Gonna call this out of scope for now.

File: src/firmware/main/app/primitives/move_primitive.c (hunk 11 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/move_primitive.c]: ah, thanks, removed here.

File: src/firmware/main/app/primitives/primitive_manager.h (hunk 12 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/primitive_manager.h]: changed as suggested

File: src/firmware/main/app/control/physbot.c [hunk 23]
```diff
@@ -0,0 +1,50 @@
+#include "firmware/main/app/control/physbot.h"
+
+#include "firmware/main/app/control/bangbang.h"
+#include "firmware/main/shared/physics.h"
+
+#define TIME_HORIZON 0.05f  // s
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/firmware/main/app/control/physbot.c:6]: I think this is out-of-scope, because it's copied exactly from the old code, and we're replacing all the controller stuff very soon.

**Comment by garethellis0** [comment]
@EvanMorcom - We definitely need to cleanup the controller and primitives, but I think that's a bit out of scope for this PR, especially given that most of that code is going to be deleted pretty soon when it's replaced with the new controller. Let me know if there's anything you strongly feel should go in this PR though!

File: src/firmware/main/app/control/bangbang.h [hunk 24]
```diff
@@ -64,38 +62,32 @@ typedef struct
...
+                                        float MaxA, float MaxV);
+void app_bangbang_prepareTrajectory(BBProfile *b, float d, float vi, float vf,
+                                    float MaxA);
 
 // meat and potatoes, computes the relavent details of the trajectory
-void PlanBBTrajectory(BBProfile *b);
+void app_bangbang_planTrajectory(BBProfile *b);
 
 // the default acceleration calculator, a typical horizon would be on
 // the time scale fo the vehicle in question say 1/2 second for our robots.
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/firmware/main/app/control/bangbang.h]: Changed as suggested and removed the bit with specific values

File: src/firmware/main/app/primitives/primitive_manager.c (hunk 17 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/primitive_manager.c]: changed as suggested

File: src/firmware/main/control/control.c (hunk 18 above)
Comment (garethellis0) [reply src/firmware/main/control/control.c:10]: We need to change a lot of these all over the place, going to say that's out-of-scope for this PR.

File: src/firmware/main/io/dr.c (hunk 19 above)
Comment (garethellis0) [reply src/firmware/main/io/dr.c:10]: Similar comment to above, out-of-scope for this PR.

File: src/firmware/main/io/dribbler.c (hunk 20 above)
Comment (garethellis0) [reply src/firmware/main/io/dribbler.c:14]: Similar comment to above, out-of-scope for this PR.

File: src/firmware/main/io/receive.c (hunk 21 above)
Comment (garethellis0) [reply src/firmware/main/io/receive.c:37]: Similar comment to above, out-of-scope for this PR.

File: src/firmware/main/main.c (hunk 22 above)
Comment (garethellis0) [reply src/firmware/main/main.c:59]: Similar comment to above, out-of-scope for this PR.

File: src/firmware/main/app/primitives/primitive_manager.c [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+struct PrimitiveManager
+{
+// The mutex that prevents multiple entries into the same primitive at the same time.
+#ifdef __arm__
+    SemaphoreHandle_t primitive_mutex;
+#elif __unix__
+    pthread_mutex_t primitive_mutex;
+#endif
+
+    // TODO: rename, this is really a pointer to the _functions_ of the current
```
Comment (EvanMorcom) [inline src/firmware/main/app/primitives/primitive_manager.c]: TODO: Rename what exactly?

File: src/firmware/main/app/primitives/shoot_primitive.c [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+#include "firmware/main/app/control/bangbang.h"
+#include "firmware/main/app/control/control.h"
+#include "firmware/main/app/control/physbot.h"
+#include "firmware/main/shared/physics.h"
+#include "firmware/main/shared/util.h"
+
+// these are set to decouple the 3 axis from each other
+// the idea is to clamp the maximum velocity and acceleration
+// so that the axes would never have to compete for resources
+#define TIME_HORIZON 0.05f  // s
```
Comment (EvanMorcom) [inline src/firmware/main/app/primitives/shoot_primitive.c:13]: Same as above. Comments and the define seem unrelated.

File: src/firmware/main/app/primitives/shoot_primitive.c [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+        float abs_factor        = fabsf(pb->min.disp) / distance_vector;
+        float minor_axis_factor = 1 - abs_factor;
+        pb->maj.accel *= minor_axis_factor;
+    }
+
+    // Doesn't work that well. Possibly because total_rot is updated too often
+    //    if (total_rot != 0) {
+    //        float rot_factor = 1 - fabsf(pb->rot.disp / total_rot);
+    //        pb->maj.accel *= rot_factor;
+    //    }
```
Comment (EvanMorcom) [inline src/firmware/main/app/primitives/shoot_primitive.c:46]: Commented out code

Remove?

File: src/firmware/main/tick.c [hunk 28]
```diff
@@ -118,6 +120,18 @@ static void normal_task(void *UNUSED(param))
         // Submit the log record, if we filled one.
         if (record)
         {
+            int current_primitive_index =
+                app_primitive_manager_getCurrentPrimitiveIndex(primitive_manager);
```
Comment (EvanMorcom) [inline src/firmware/main/tick.c:124]: getCurrentPrimitiveIndex() returns an unsigned int.

Probably won't cause any issues, but maybe?

File: src/software/backend/simulation/simulator_robot.h [hunk 29]
```diff
@@ -218,6 +218,22 @@ class SimulatorRobotSingleton
...
+    /**
+     * Sets the motor to coast (spin freely)
+     */
+    static void coastMotorBackLeft();
+    static void coastMotorBackRight();
+    static void coastMotorFrontLeft();
+    static void coastMotorFrontRight();
+
+    /**
+     * Sets the motor to brake (act against the current direction of rotation)
```
Comment (EvanMorcom) [inline src/software/backend/simulation/simulator_robot.h:230]: We might need to elaborate on what 'braking' means.

It is trying to achieve zero velocity?

File: src/firmware/main/app/primitives/move.c [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+
+#include "firmware/main/app/control/bangbang.h"
+#include "firmware/main/app/control/control.h"
+#include "firmware/main/app/control/physbot.h"
+#include "firmware/main/shared/physics.h"
+#include "firmware/main/shared/util.h"
+#include "shared/constants.h"
+#include "shared/robot_constants.h"
+
+// TODO: all primitive functions need prefixing
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/firmware/main/app/primitives/move.c]: So we should definitely prefix all primitive functions with `app_`, but I'm going to hold off on doing this for now as that takes a fair amount of time and a lot of primitives are going to be deleted soon. Just removed for now.

File: src/firmware/main/app/control/physbot.h (hunk 9 above)
Comment (garethellis0) [reply src/firmware/main/app/control/physbot.h]: renamed

File: src/firmware/main/app/primitives/primitive_manager.c (hunk 25 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/primitive_manager.c]: Ah, so I renamed `current_primitive` to `current_primitive_functions` because I want it to be clear that this member does not hold the _state_ of the primitive.

File: src/firmware/main/app/primitives/shoot_primitive.c (hunk 26 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/shoot_primitive.c:13]: Easy fix, so I'll change this one. Nuked the comment.

File: src/firmware/main/app/primitives/shoot_primitive.c (hunk 27 above)
Comment (garethellis0) [reply src/firmware/main/app/primitives/shoot_primitive.c:46]: There's a bunch of this in many primitives, calling out-of-scope for now.

File: src/firmware/main/tick.c (hunk 28 above)
Comment (garethellis0) [reply src/firmware/main/tick.c:124]: _and then_ we cast it to a `uint8_t` below..... 

So yeah, changed to `unsigned int`.

File: src/software/backend/simulation/simulator_robot.h (hunk 29 above)
Comment (garethellis0) [reply src/software/backend/simulation/simulator_robot.h:230]: TBD? I don't know what the expected behaviour from the robot side is, but I suspect it's just not free-wheeling, so the back-emf is all that's acting against it. @MathewMacDougall @hannahvsawiuk ?

Comment (EvanMorcom) [reply src/software/backend/simulation/simulator_robot.h:230]: I'm just wondering if it is breaking, buy how much? And what determines that?

Comment (garethellis0) [reply src/software/backend/simulation/simulator_robot.h:230]: That'll depend on the motor constants (speed and torque constants). I'm not sure if we want to strictly define that here or not, just because it will be hardware dependent. (This is all assuming my guess for what this does is correct)

**Review by MathewMacDougall — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]
LGTM

File: src/software/backend/simulation/simulator_robot.h (hunk 29 above)
Comment (EvanMorcom) [reply src/software/backend/simulation/simulator_robot.h:230]: Yeah, our maximums are based on the motor parameters. What I mean is, is it trying to brake at 100% effort, or 20% ect

Comment (EvanMorcom) [reply src/software/backend/simulation/simulator_robot.h:230]: But out of scope

**Review by EvanMorcom — APPROVED** [review]
🥇

**Review by MathewMacDougall — APPROVED** [review]

**Comment by MathewMacDougall** [comment]
@LiCody can you re-review?

**Comment by LiCody** [comment]
LGTM

**Review by LiCody — APPROVED** [review]
